Feature #213: Optional feature to prevent any opening movement of a cover while still allowing closing movements, could be used to force drying of a cover

FrankTub opened the pull request:
I would like some feedback on my implementation direction. 

Some trade-offs I made:

- It has the same trade off as the other force features, the state if the option was not on/true can not be reproduced when this changes back to false
- Still in doubt if below logic should be applied to all actions where the action `cover.set_cover_position` or `cover.open/close_cover` is called. ~~I currently think did not do this for the "close" situation but this might be a mistake.~~
- if a cover is not moved because of this option, the tilt change is also not applied. I think I made it clear enough in the description that this is the behavior and as far as I can imagine this will most likely only be applied for covers that are outside and thus do not have a tilt option.
- If another force feature is applied it will still take a look at the new feature `auto_prevent_opening_force` if this force feature is allowed to happen. To me it makes the most sense if someone is using multiple force features
- I placed this feature under force since it behaves differently compared to `prevent_*` due to the fact that the state is not reproduced when this option changes to false/off.

~~I'm still testing this a bit, so might be best to wait for my confirmation that everything works OK in my case at least~~ All tests look OK

Commit:
First try to implement an option to prevent opening of the cover completely, while still maintaining other functionalities

Commit:
Clarify new option naming and align it with other settings

Commit:
Repositioned condition check if we should allow the move of the cover

Commit:
Hufterproof prevent opening force feature in case the close position is higher than the current position of the cover. This should be an edge case

hvorragend commented:
I really appreciate it when someone helps out here. Many thanks for that. But I'm a little unsure about the merge. 

First of all, all existing force features are there to trigger a certain state. This does not allow the mode or behavior of the automation to be changed. I have added the feature flags under Automation Options for this purpose. Here you can control very deficient little things. 
Okay, admittedly. Until now, you could not specify an entity here. 

I would prefer that up there. 
Unfortunately, the code base has changed completely today. A merge would not be possible at the moment anyway. I'm sorry, but my local repository was already further along.

FrankTub commented:
Thanks for your feedback! 

The thing I'm still a bit worried on is if we add it to the prevent list will this behave in the same way as the other prevent options or do I need to make it clear in the description or something that the behavior might be different compared to the rest of the list?

No problem for me to rewrite it to the new situation. Will do so once you come back to the above question.

I was also considering yaml anchors in this PR but decided not to do so since you have not used it elsewhere. Great to see that you implemented this now so that it becomes even better maintainable and readable!

FrankTub commented:
I agree that complexity always comes at a price of the maintainability. So perhaps I should elaborate a bit more on what I want to achieve and what I considered as alternatives.

## Setup 

I have 3 roller shutters and 1 screen with 2 on the south side and 2 on the north side. 

For the 2 on the north side the automation is relatively limited in terms of the options I use. I use generic opening and closing of the cover and in addition use the sun elevation option and intend to start using the residence setting when I got proper sensors to depend it on.

The 2 on the south side are also straightforward. In addition to the above options I also use sun shading functionality.

I also configured the working day sensor to distinguish between late and early drives of the automation. In addition I use a helper in all 4 automations.

## Use case 

When it is raining my covers get wet since they are placed at the outside of the windows. Especially for the screen it is important that it does not move to an open position to prevent mold. But also for the roller shutters it is advised by the manufacturer to not open the roller shutters when they are wet. This will enhance life expectancy for these devices. 

## Ideal solution

When it is raining or if a cover still needs to dry **after** the rain has stopped it should be templates in a binary sensor. This binary sensor should be used as an input for a new feature.

### Not closed 

When my covers are in a not closed position (i.e. open/ventilation/sun shading (unlikely)) and apparently we prefer to have light from the outside in our house and it will start raining nothing needs to happen directly.

If it keeps on raining and the cover should move the a more closed position this is no problem and this should still happen. This could be a closed position, but also perhaps a sun shading/ventilation position.

If it stops raining and the cover was still not closed, then it should move to the position it otherwise should have been in without this feature. Given that closing movements were still allowed this allows only movement from sun shading /ventilation / open position.

### Closed

When it stops raining and it is still before the time that the cover should move to an upward position then it should stay as it is.

When it keeps on raining till the point where normally the cover would have opened, it should stay as it was -> closed.

When it stops raining after the point where the cover normally would have opened if this feature wasn't there, then it should move to the position where it would have been without this feature. I think this could also be the case that it could move to a ventilation or sun shading position aside from open. It depends on the entity and your drying period what the right position might be.

## Alternatives considered

- Having a custom automation that sets the above mentioned binary sensor and disable the automation through a force feature in cca. A need to reproduce the state the cca automation would have been in without this binary sensor. This is very hard given that you need to rebuild all logic in cca to determine proper location.
- Using one of the other force features. I don't want to force a specific state of it is raining. For example I don't want to always open a cover to prevent it from getting wet, this is not useful for example in the middle of the night. I also don't want to make it unnecessary dark in rooms simply always closing the covers when it is raining.

Perhaps I'm missing something obvious that is already there that I could use. Please let me.know if that is the case

Ps: this PR does not adhere to my description of the ideal solution. It does not add a trigger when this entity moves from on to off, which is probably what would be required.

Commit:
Refactored to new yaml anchors and made code even more DRY

Commit:
Move new feature to prevent options instead of force

Commit:
Add a trigger that when opening movements are allowed again the state is reproduced as it was expected to be otherwise

Commit:
Remove blank line introduction

Commit:
Minor fix for prevent_opening_movement_disabled

Commit:
test

Commit:
Bugfix logic helper update and coer move actions

Commit:
Target position can be empty??

Commit:
Bugfix reproduce ventilation mode if window was opened while prevent_opening_movement was on and this feature turned back off while cover was still closed

FrankTub commented:
@hvorragend I refactored and tested the new code and I'm really happy about it the way it is working for me. I introduced a new anchor and hope this is something you agree on 🙏 

This new option `prevent_opening_movement` does the following:

- When this entity of `prevent_opening_movement` changes from off to on, then nothing happens right away
- When this entity of `prevent_opening_movement` changes from on to off, it will trigger the automation.
  - When the exepected state of the cover would have been open, it will change to open (Verified and tested)
  - When the expected state of the cover is closed, then it will stay closed (Verified and tested)
  - When the current state is closed and the expected state is ventilation, then it will change to ventilation position (Verified and tested)
  - When the current state is closed or ventilation and the expected state is shading (and the position is higher than closed/ventilation), I guess I made the required changes but  **I have not tested this in my setup**.
  - When the current state is shading position and shading has ended it should change to the new expected situation (that his higher than the shading position), I think I made the required changes, but **I have not tested this and unsure how to do this.**

Also verified the following, `prevent_opening_movement = on`:
- [x] all downward movements are executed as they normally would.
- [x] upward movements as open/ventilation position are not occuring
When `prevent_opening_movement = off`:
- [x] all downward movements are executed as they normally would.
- [x] All upward movements are happening as they normally would

Let me know what you think and if anything needs to be tweaked. Also open for renaming some of these parts of the code where ever necessary. Otherwise also happy to just use my custom version of the code if you don't agree on it.

hvorragend commented:
Thank you very much for your effort. You're making a real effort and testing it too. I think that's really good.

I'm still unsure about this change. But I understand it very well. So you don't want the roller blind to open any further. And that across the entire code in all features. And since you have to check the positions, you can't use the existing conditions for this. That's the reason, isn't it?

Everything understandable up to this point.

I'm just annoyed that you have combined my three YAML anchors (`*cover_move_action`, `*tilt_move_action`, `*helper_update`) into "`*cover_move_and_tilt_and_update_helper`".
This makes it much harder to maintain the code and the flexibility is lost. But I don't have a better idea right now.

Although I would try to incorporate this into the existing YAML anchors. I'll take a look at it.

I also find this nesting absolutely unfortunate. At first glance, I would say that it should be much easier. I think I can see a lot of redundancies.


        - if:
            - or:
                - and:
                    - or:
                        - "{{ prevent_opening_movement_disabled }}"
                        - "{{ not prevent_opening_movement_disabled and (current_position >= target_position | default(101)) }}"
                    - or:
                        - "{{ should_cover_move_and_tilt is not defined }}"
                        - "{{ should_cover_move_and_tilt is defined and should_cover_move_and_tilt }}"
                - "{{ should_cover_move_and_tilt is defined }}"

FrankTub commented:
Now I tried to simplify I realised I also had another reason for combining the three YAML anchors into a single anchor. 

> The trade off is that some parts of the code ALWAYS execute `*helper_update`, while `*cover_move_action` and `*tilt_move_action` are still conditional.

This is still true. The hard part here is that if the part where you call one of the anchors, the part where you called it from is not aware of if actually executed that step. Now I realise that that was the reason why I decided to combine the YAML anchors in the first place. So if the cover is not moved, it does not makes sense to update the helper. However, sometimes you still need to update the helper, even if the cover is not moved (not sure why though). &rarr; You can't simply place the same logic in `helper_update` as you would in `cover_move_action`. I'm a bit out of inspiration how I can solve this without an ugly if statement and a new variable in the places where you always want to update the helper independent if you move the cover. What is your take on it @hvorragend ?

hvorragend commented:
We replied at the same time. My last answer does not refer to your last post from just now. But to the one before that.  😀

FrankTub commented:
What I have so far is https://github.com/FrankTub/ha-blueprints/blob/feature/feedback-prevent-opening/blueprints/automation/cover_control_automation.yaml. But it needs work on the helper cause that is really not correct in this commit. Would be really appreciated if you could help out here 🙏

hvorragend commented:
Please give me a little more time.

FrankTub commented:
@hvorragend do you expect me to rebase it given that there is a new version available?
